src/mqtt_interface.rs

```diff
@@ -0,0 +1,258 @@
+use super::StringSet;
+use core::fmt::Write;
+use heapless::{consts, String};
+use minimq::{
+    embedded_nal::{IpAddr, TcpStack},
+    MqttClient, Property, QoS,
+};
+
+#[derive(Debug)]
+pub enum Error<E: core::fmt::Debug> {
+    /// The provided device ID is too long.
+    IdTooLong,
+
+    /// MQTT encountered an internal error.
+    Mqtt(minimq::Error<E>),
+
+    /// The network stack encountered an error.
+    Network(E),
+}
+
+impl<E: core::fmt::Debug> From<minimq::Error<E>> for Error<E> {
+    fn from(err: minimq::Error<E>) -> Self {
+        match err {
+            minimq::Error::Network(err) => Error::Network(err),
+            other => Error::Mqtt(other),
+        }
+    }
+}
+
+/// An action that applications should act upon.
+#[derive(PartialEq)]
+pub enum Action {
+    /// There is nothing to do except continue normal execution.
+    Continue,
+
+    /// The settings are being commit to memory. The application should take steps to make staged
+    /// settings active.
+    CommitSettings,
+}
+
+// Generate an MQTT topic of the form `<device_id>/<topic>`.
+//
+// # Returns
+// The string - otherwise, an error indicating the generated string was too long.
+fn generate_topic<'a, 'b>(device_id: &'a str, topic: &'b str) -> Result<String<consts::U128>, ()> {
+    let mut string: String<consts::U128> = String::new();
+    write!(&mut string, "{}/{}", device_id, topic).or(Err(()))?;
+    Ok(string)
+}
+
+/// An interface for managing MQTT settings.
+pub struct MqttInterface<T, S>
+where
+    T: StringSet,
+    S: TcpStack,
+{
+    // TODO: Allow the user to specify buffer size.
+    client: Option<MqttClient<minimq::consts::U256, S>>,
```

Comment:
It seems like `client` is never `None` except within the body of each function, so why wrap in an `Option`? Is it to protect against the possibility of two different functions being called at the same time? Basically panic if the user uses the library wrong?

Reply:
This is actually to get around rust's ownership rules. The issue here is that `minimq::MqttClient` takes a `FnMut` closure to process incoming messages. However, that results in us mutably borrowing the `self.client`. This means that we aren't allowed to then use `self` mutably _within_ that closure. The solution is to store the `client` in an `Option` and move it in and out of `self` as needed - I'm open to other ideas though.

Reply:
I see....Hmm I was thinking a Cell might help there, but that would look really similar to what you already have.